ansible/roles/test/files/ptftests/hash_test.py

```diff
@@ -98,63 +74,23 @@ def check_hash(self, src_ip_range, dst_ip_range, ipv4=True):
-
-        # step 2: check randomizing destination ip
-        if self.test_hash_dstip:
-            hit_count_map.clear()
-            for i in range(0, self.BALANCING_TEST_TIMES):
-                dst_ip = dst_ip_interval.get_random_ip()
-                (matched_index, _) = self.check_ip_route(
-                        in_port, src_port, dst_port, src_ip, dst_ip, exp_port_list, ipv4)
-                hit_count_map[matched_index] = hit_count_map.get(matched_index, 0) + 1
-            logging.info("hit count map: {}".format(hit_count_map))
-            self.check_balancing(next_hop.get_next_hop(), hit_count_map)
-
-        # step 3: check randomizing l3 source port
-        if self.test_hash_srcport:
-            hit_count_map.clear()
-            for i in range(0, self.BALANCING_TEST_TIMES):
-                src_port = random.randint(0, 65535)
-                (matched_index, _) = self.check_ip_route(
-                        in_port, src_port, dst_port, src_ip, dst_ip, exp_port_list, ipv4)
-                hit_count_map[matched_index] = hit_count_map.get(matched_index, 0) + 1
-            logging.info("hit count map: {}".format(hit_count_map))
-            self.check_balancing(next_hop.get_next_hop(), hit_count_map)
-
-        # step 4: check randomizing l4 destination port
-        if self.test_hash_dstport:
-            hit_count_map.clear()
-            for i in range(0, self.BALANCING_TEST_TIMES):
-                dst_port = random.randint(0, 65535)
-                (matched_index, _) = self.check_ip_route(
-                        in_port, src_port, dst_port, src_ip, dst_ip, exp_port_list, ipv4)
-                hit_count_map[matched_index] = hit_count_map.get(matched_index, 0) + 1
-            logging.info("hit count map: {}".format(hit_count_map))
-            self.check_balancing(exp_port_list, hit_count_map)
-
-        # step 5: check randomizing in port
-        # TODO
-
-    def check_ip_route(self, in_port, sport, dport, src_ip_addr, dst_ip_addr,
-                       dst_port_list, ipv4=True):
-        if ipv4:
-            (matched_index, received) = self.check_ipv4_route(in_port, sport, dport,
-                    src_ip_addr, dst_ip_addr, dst_port_list)
+        for _ in range(0, self.BALANCING_TEST_TIMES):
+            src_port = random.choice([port for port in self.src_ports if port not in exp_port_list]) if hash_key == "ingress-port" else in_port
+            logging.info("src_port: {}".format(src_port))
+            (matched_index, _) = self.check_ip_route(hash_key, src_port, dst_ip, exp_port_list)
+            hit_count_map[matched_index] = hit_count_map.get(matched_index, 0) + 1
+        logging.info("hit count map: {}".format(hit_count_map))
+
+        # The sample is too little for hash_key vlan-id and ingress-port, check it loose
+        check_balancing_mode = 'loose' if hash_key in ['vlan-id', 'ingress-port'] else 'strict'
+        self.check_balancing(next_hop.get_next_hop(), hit_count_map, check_balancing_mode)
```

Comment:
i am not sure why we introduce loose mode, it seems defeating the purpose of the test.

Reply:
Load balance is established in the case of a large number of samples. When the sample size is too small, the result may be hashed to the same destination, and the result is load balance not very well. In our hash test, the number of vlan-id and ingress-port was small, so i add them to the loose mode.